lib/DxilValidation/DxilValidation.cpp

```diff
@@ -970,6 +970,226 @@ static void ValidateImmOperandForMathDxilOp(CallInst *CI, DXIL::OpCode Opcode,
+  case DXIL::OpCode::MatVecMulAdd:
+    return CI->getOperand(DXIL::OperandIndex::kMatVecMulAddIsOutputUnsignedIdx);
+
+  default:
+    DXASSERT_NOMSG(false);
+    return nullptr;
+  }
+}
+
+static void ValidateImmOperandsForMatVecOps(CallInst *CI, DXIL::OpCode OpCode,
+                                            ValidationContext &ValCtx) {
+
+  // Check Common operands
+  llvm::Value *IsInputUnsigned =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulIsInputUnsignedIdx);
+  llvm::Value *InputInterpretation =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulInputInterpretationIdx);
+  llvm::Value *MatrixInterpretation =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixInterpretationIdx);
+  llvm::Value *MatrixM =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixMIdx);
+  llvm::Value *MatrixK =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixKIdx);
+  llvm::Value *MatrixLayout =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixLayoutIdx);
+  llvm::Value *MatrixTranspose =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixTransposeIdx);
+  llvm::Value *IsOutputUnsigned = getMatVecOpIsOutputUnsigned(CI, OpCode);
+
+  ConstantInt *IsInputUnsignedConst =
+      dyn_cast<llvm::ConstantInt>(IsInputUnsigned);
+  if (!IsInputUnsignedConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    return;
+  }
+
+  ConstantInt *IsOutputUnsignedConst =
+      dyn_cast<llvm::ConstantInt>(IsOutputUnsigned);
+  if (!IsOutputUnsignedConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    return;
+  }
+
+  ConstantInt *II = dyn_cast<ConstantInt>(InputInterpretation);
+  ConstantInt *MI = dyn_cast<ConstantInt>(MatrixInterpretation);
+  if (!II || !MI) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+    return;
+  }
+
+  // Check if InputInterpretation and MatrixInterpretation are valid
+  auto IIValue = II->getLimitedValue();
+  if (!CheckFromRegisterInterpretations(IIValue)) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInvalidRegisterInterpValue);
+    return;
+  }
+
+  auto MIValue = MI->getLimitedValue();
+  if (!CheckInMemoryInterpretations(MIValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMemoryInterpValue);
+    return;
+  }
+
+  ConstantInt *MatrixTransposeConst = dyn_cast<ConstantInt>(MatrixTranspose);
+  ConstantInt *MatrixLayoutConst = dyn_cast<ConstantInt>(MatrixLayout);
+  if (!llvm::isa<llvm::Constant>(MatrixM) ||
+      !llvm::isa<llvm::Constant>(MatrixK) || !MatrixLayoutConst ||
+      !MatrixTransposeConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgMatrixShapeParamsAreConst);
+    return;
+  }
+
+  auto MLValue = MatrixLayoutConst->getLimitedValue();
```

Comment:
nit: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
```suggestion
  uint64_t MLValue = MatrixLayoutConst->getLimitedValue();
```